### ISSUE TITLE
acme.sh --home directory question

### ISSUE THREAD

**Issue by tackynugget** [lippertmarkus/synology-le-dns-auto-renew#2]
Hi,

Thank you for your work here: https://lippertmarkus.com/2020/03/14/synology-le-dns-auto-renew/

I am just getting started with these sorts of modifications and was able to follow your guides very easily. In fact, so easily that I didn't catch the complications that could arise from setting the home parameter to something other than the default `~/acme.sh` directory.

The first snag came when I wanted to enable `--auto-upgrade`
The second snag came when I wanted to use acme.sh to convert my certs `--to-pkcs12`

I figured out the `--home /usr/local/share/acme.sh/` parameter in all of the acme.sh commands here was what redirected the action to the `/usr/local/share/acme.sh/` rather than the default `~/acme.sh/` folder.

However, I am curious about the security implications of storing certs in general and unencrypted keys in particular in a shared location with these permissions:
`drwxrwxrwx certadmin users`

versus the home directory with these permissions
`drwx--x--x+ certadmin users`

I get the idea of putting the certs in a shared space, but it seems like **_IF_** the `certadmin` account is ONLY used for Let's encrypt certificate renewals, it would make more sense to change **_that_** home folder's group permissions to grant access to the accounts that need to access certs rather than put certs and keys in a place where all users can access it.

Thoughts?

**Comment by lippertmarkus** [comment]
Hey, thanks a lot for the feedback!

When thinking again about that I wonder if it's even necessary to allow other users to have access to the certificates as they are retrieved by `certadmin` and uploaded via the API by `certadmin` and that should be it. Could you try doing the setup within the home directory of `certadmin` to see if that works without a problem?

**Comment by tackynugget** [comment]
Yes @lippertmarkus. I'd be happy to test. However, it may be a while before I can. I'll be sure to report back ASAP.

**Comment by lippertmarkus** [comment]
sure, take the time. Thanks!

**Comment by tackynugget** [comment]
@lippertmarkus I started to test this. 

How important is it that the `certadmin` account be an administrator? I'm finding that members of the `Administrators` group cannot have custom permissions to the `/homes` share. I was trying to scope it so that my `certadmin` account would only have access to its own `~/` directory and not to any of the other home directories. 

The only way I can scope custom permissions to the `/homes` share is by removing the `certadmin` account from the `Administrators` group.

**Comment by lippertmarkus** [comment]
@OverengineeredNetwork Administrator rights are needed for uploading the certs via the DSM API so that could get difficult

**Comment by tackynugget** [comment]
Thank you for that reminder about the importance of the administrators role in importing the cert into the Synology.

So, here are my findings:

I opted **not** to pursue installing acme.sh in the default `~/acme.sh` directory because it would require granting the `certadmin` service account as well as any `otherservices`  service accounts read/write access to the `/var/services/homes` directory, which violates my application of the principle of least privilege. However, based on my testing below, I am sure that granting `certadmin` read/write access to `homes` and some additional permission modifications to its home directory for any `otherservices` service accounts would work if someone was interested in this and was comfortable with the risk.

As I see it, the better approach is to maintain the installation in `/usr/local/share/acme.sh` as you have it, keep its permissions locked down (since it holds the certificate key in plaintext), and to create a separate directory and user group for sharing the exported, password-protected pfx file (see below).

As for my initial question, both tasks (enabling auto-upgrade and PFX conversion) can be accomplished by adding the `--home /usr/local/share/acme.sh/` parameter to their respective commands. If scripting this, I'd recommend setting an acme_path variable for readability:
`acme_path=/usr/local/share/acme.sh`

**Auto-upgrade**
`$acme_path/acme.sh --upgrade --auto-upgrade --home $acme_path`

**PFX conversion**
`$acme_path/acme.sh --toPkcs -d $domain_name --password $p12cert_password --home $acme_path`

The PFX conversion absolutely requires the `administrators` group have rwx permissions to the `/usr/local/share/acme.sh` directory because, as far as I can tell, we cannot specify a different destination for the PFX output file. However, I was able to script a separate `copy` (`move` works too) operation to a folder where other services could access the password-protected PFX cert (i.e. `/usr/local/share/certshare`) by creating a new shared group between the `certadmin` and the `otherservices` user.

SSH'd into my Synology as `certadmin`
`ssh certadmin@domain.tld`

Elevated to root
`sudo -i`

Created a `certread` group and made the `certadmin` and `otherservices` users members
`synogroup --add certread certadmin otherservices`

Verified member of changes to the two users
`synouser --get certadmin`
`synouser --get otherservices`

Created a new certificates folder in /usr/local/share/certshare
`mkdir /usr/local/share/certshare`

Set the initial permissions (the certread group automatically received xr rights)
`chown -R certadmin:certread /usr/local/share/certshare/`

Copied the pfx cert from the protected area into the shared area
`cp /usr/local/share/acme.sh/domain.tld/domain.tld.pfx /usr/local/share/certshare/domain.tld.pfx`

Logged out of my SSH session and then logged back in because the newly created `certread` group permission would not take effect and it needs to be in effect to be able to set the permissions for the copied file to the new group.
`chown -R certadmin:certread /usr/local/share/certshare/domain.tld.pfx`
